### PR TITLE
AArch64: save FPU context on ctx_switch, restore on user_exc_leave.

### PR DESCRIPTION
**Pull request by pj1031999**
Port #928 for _AArch64_.

### CODE REVIEW

**Comment by cahirwpz** [comment]
Could you move MIPS implementation to separate PR?

**Comment by pj1031999** [comment]
@cahirwpz see #928.

File: sys/aarch64/evec.S [hunk 1]
```diff
@@ -134,6 +162,35 @@ ENTRY(handle_user_trap)
...
+        cmp     w2, wzr
+        beq     .skip_fpu_restore
+
+        /* clear TDP_FPUCTXSAVED flag */
+        and     w2, w3, ~TDP_FPUCTXSAVED
+        str     w2, [x1, #TD_PFLAGS]
+
+        /* restore FPU context */
+        ldr     x1, [x1, #TD_UCTX]
+        load_fpu_ctx x1, x2, w2 /* 2nd and 3rd argument must be the same */
```
Comment (franciscozdo) [inline sys/aarch64/evec.S:191]: Here I would write sth like '2nd and 3rd argument must be register and its lower half'.

File: sys/aarch64/switch.S [hunk 2]
```diff
@@ -47,6 +75,32 @@ ENTRY(ctx_switch)
...
+
+        /* enable FPU and save context */
+        /* thread_t::tdp_flags is a volatile unsigned - use 32-bit */
+        mrs     x2, cpacr_el1
+        and     x2, x2, ~CPACR_FPEN_MASK
+        orr     x2, x2, CPACR_FPEN_TRAP_NONE
+        msr     cpacr_el1, x2
+
+        ldr     x2, [x0, #TD_UCTX]
+        save_fpu_ctx x2, x3, w3
```
Comment (franciscozdo) [inline sys/aarch64/switch.S:95]: Here you should also add a comment about 2nd and 3rd argument as you did before.

**Review by franciscozdo — CHANGES_REQUESTED** [review]
I suggest rewriting comments then code will be more reader friendly. Also I don't understand one use of `TDP_FPUCTXSAVED` flag.

File: sys/aarch64/evec.S [hunk 3]
```diff
@@ -106,6 +106,34 @@
 .endif
 .endm
 
+/* 
+ * WARNING!
+ * xtmp & wtmp MUST BE the same register - fpcr and fpsr are 32-bit registers
```
[the remark this answers is not in the document]
Comment (pj1031999) [reply sys/aarch64/evec.S]: done

File: sys/aarch64/switch.S [hunk 4]
```diff
@@ -19,6 +19,34 @@
         stp      x8,  x9, [sp, #CTX_X8]
 .endm
 
+/* 
+ * WARNING!
+ * xtmp & wtmp MUST BE the same register - fpcr and fpsr are 32-bit registers
```
[the remark this answers is not in the document]
Comment (pj1031999) [reply sys/aarch64/switch.S]: done

File: sys/aarch64/switch.S [hunk 5]
```diff
@@ -47,6 +75,32 @@ ENTRY(ctx_switch)
 
         # save context of @from thread
 .ctx_save:
+        ldr     w3, [x0, #TD_PFLAGS]
+        and     w2, w3, #TDP_FPUINUSE|TDP_FPUCTXSAVED
```
[the remark this answers is not in the document]
Comment (pj1031999) [reply sys/aarch64/switch.S:78]: I use that flag.

**Review by cahirwpz — APPROVED** [review]
LGTM

**Pull request merged**